selene-lib/src/rules/mismatched_arg_count.rs

```diff
@@ -0,0 +1,425 @@
+            // there is no upper bound since foo(a, b, c, d) is valid - therefore any amount of arguments provided is valid
+            ParameterCount::Minimum(_) => true,
+            // Any amount of arguments could be provided
+            ParameterCount::Variable => true,
+        }
+    }
+
+    pub fn to_message(&self, provided: PassedArgumentCount) -> String {
+        match self {
+            ParameterCount::Fixed(required) => {
+                format!("{} arguments but {} were supplied", required, provided)
+            }
+            ParameterCount::Minimum(required) => format!(
+                "atleast {} arguments but {} were supplied",
+                required, provided
+            ),
+            ParameterCount::Variable => "a variable amount of arguments".to_owned(),
+        }
+    }
+}
+impl Display for ParameterCount {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ParameterCount::Fixed(required) => write!(f, "expected {} arguments", required),
+            ParameterCount::Minimum(required) => {
+                write!(f, "expected atleast {} arguments", required)
+            }
+            ParameterCount::Variable => write!(f, "expected any number of arguments"),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum PassedArgumentCount {
+    /// Passed a fixed amount of arguments, such as foo(a, b, c) or foo(a, call(), c) or foo(a, ..., c)
+    Fixed(usize),
+    /// Passed a variable of arguments - but we know the lower bound: e.g. foo(a, b, call()) or foo(a, b, ...)
+    Variable(usize),
+}
+impl Display for PassedArgumentCount {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            PassedArgumentCount::Fixed(amount) => write!(f, "{} arguments", amount),
+            PassedArgumentCount::Variable(amount) => write!(f, "atleast {} arguments", amount),
+        }
+    }
+}
+
+fn function_call_argument_count(function_args: &ast::FunctionArgs) -> PassedArgumentCount {
+    match function_args {
+        ast::FunctionArgs::Parentheses { arguments, .. } => {
+            // We need to be wary of items with side effects, such as function calls or ... being the last argument passed
+            // e.g. foo(a, b, call()) or foo(a, b, ...) - we don't know how many arguments were passed.
+            // However, if the call is NOT the last argument, as per Lua semantics, it is only classed as one argument, and no side effects occur
+            // e.g. foo(a, call(), b) or foo(a, ..., c)
+
+            let mut passed_argument_count = 0;
+
+            for argument in arguments.pairs() {
+                match argument {
+                    ast::punctuated::Pair::Punctuated(_, _) => passed_argument_count += 1,
+                    ast::punctuated::Pair::End(expression) => {
+                        passed_argument_count += 1;
```

Comment:
Both branches have this, so just do `if let ast:punctuated::Pair::End(expression) = argument` instead.